File: src/Atlasd/Battlenet/Protocols/BNFTP/File.cs
using Atlasd.Daemon;
using System;
using System.IO;

namespace Atlasd.Battlenet.Protocols.BNFTP
{
    class File : IDisposable
    {
        public string BNFTPPath { get => Settings.GetString(new string[] { "bnftp", "root" }, null); }
        public bool Exists { get => GetFileInfo().Exists; }
        public bool IsDirectory { get => GetFileInfo().Attributes.HasFlag(FileAttributes.Directory); }
        public string Name { get; private set; }
        public DateTime LastAccessTime { get => System.IO.File.GetLastAccessTime(System.IO.Path.Combine(BNFTPPath, Name)); }
        public DateTime LastAccessTimeUtc { get => System.IO.File.GetLastAccessTimeUtc(System.IO.Path.Combine(BNFTPPath, Name)); }
        public DateTime LastWriteTime { get => System.IO.File.GetLastWriteTime(System.IO.Path.Combine(BNFTPPath, Name)); }
        public DateTime LastWriteTimeUtc { get => System.IO.File.GetLastWriteTimeUtc(System.IO.Path.Combine(BNFTPPath, Name)); }
        public long Length { get => GetFileInfo().Length; }
        public string Path { get => System.IO.Path.GetFullPath(System.IO.Path.Combine(BNFTPPath, Name)); }
        public StreamReader StreamReader { get; private set; } = null;

        public File(string filename)
        {
            Name = filename;
        }

        /**
         * <remarks>Closes any internally open references. Call this before dropping all references to this object.</remarks>
         */
        public void Close()
        {
            CloseStream(); // close the StreamReader if it's open.
        }

        /**
         * <remarks>Closes the StreamReader property if it is open.</remarks>
         */
        public void CloseStream()
        {
            if (StreamReader != null)
            {
                StreamReader.Close();
            }
        }

        /**
         * <remarks>This is part of the .NET IDisposable class interface. User code should use the Close() method instead.</remarks>
         */
        public void Dispose() /* part of IDisposable */
        {
            // call our own Close() method.
            Close();
        }

        /**
         * <remarks>Retrieves a System.IO.FileInfo object given the current values of the BNFTPPath and Name properties.</remarks>
         * <param name="ignoreLimits">Set to False to check that the target file does not leave BNFTP root, is a file, and it exists. If a check fails, the entire method stops, prints to Logging, and returns null. Set to True to ignore these checks instead and attempt creating a FileInfo object anyway.</param>
         * <returns>The System.IO.FileInfo object if successful, null otherwise.</returns>
         */
        public FileInfo GetFileInfo(bool ignoreLimits = false)
        {
            var rootStr = System.IO.Path.GetFullPath(BNFTPPath);
            var pathStr = System.IO.Path.GetFullPath(System.IO.Path.Combine(rootStr, Name));

            if (!ignoreLimits && (pathStr.Length < rootStr.Length || pathStr.Substring(0, rootStr.Length) != rootStr))
            {
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info; path would leave BNFTP root directory");
                return null;
            }

            FileInfo fileinfo;
            try
            {
                fileinfo = new FileInfo(System.IO.Path.Combine(BNFTPPath, Name));
            }
            catch (Exception ex)
            {
                if (!(ex is UnauthorizedAccessException || ex is PathTooLongException || ex is NotSupportedException)) throw;
                return null;
            }

            if (fileinfo == null)
            {
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; null FileInfo object");
                return null;
            }

            if (!ignoreLimits && !fileinfo.Exists)
            {
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; file not found");
                return null;
            }

            if (!ignoreLimits && fileinfo.Attributes.HasFlag(FileAttributes.Directory))
            {
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; path pointed to a directory");
                return null;
            }

            return fileinfo;
        }

        /**
         * <remarks>Opens a StreamReader object then assigns it to the StreamReader property, and returns true if successful; returns false otherwise.</remarks>
         * <returns>True if the StreamReader was opened and assigned to File.StreamReader, False otherwise; if False, File.StreamReader will be null.</returns>
         */
        public bool OpenStream()
        {
            if (StreamReader != null)
            {
                StreamReader.Close();
            }
            StreamReader = null;

            var fileinfo = GetFileInfo();
            if (fileinfo == null) return false;

            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.BNFTP, $"Opening read stream for [{fileinfo.FullName}]...");
            StreamReader = new StreamReader(fileinfo.FullName);

            return true;
        }
    }
}

File: src/AtlasV/Battlenet/Protocols/BNFTP/File.vb
Imports AtlasV.Daemon
Imports System
Imports System.IO

Namespace AtlasV.Battlenet.Protocols.BNFTP
    Class File
        Implements IDisposable

#Disable Warning CA1822 ' Mark members as static
        Public ReadOnly Property BNFTPPath As String
#Enable Warning CA1822 ' Mark members as static
            Get
                Return Settings.GetString(New String() {"bnftp", "root"}, Nothing)
            End Get
        End Property

        Public ReadOnly Property Exists As Boolean
            Get
                Return GetFileInfo().Exists
            End Get
        End Property

        Public ReadOnly Property IsDirectory As Boolean
            Get
                Return GetFileInfo().Attributes.HasFlag(FileAttributes.Directory)
            End Get
        End Property

        Public Property Name As String

        Public ReadOnly Property LastAccessTime As DateTime
            Get
                Return System.IO.File.GetLastAccessTime(System.IO.Path.Combine(BNFTPPath, Name))
            End Get
        End Property

        Public ReadOnly Property LastAccessTimeUtc As DateTime
            Get
                Return System.IO.File.GetLastAccessTimeUtc(System.IO.Path.Combine(BNFTPPath, Name))
            End Get
        End Property

        Public ReadOnly Property Length As Long
            Get
                Return GetFileInfo().Length
            End Get
        End Property

        Public ReadOnly Property Path As String
            Get
                Return System.IO.Path.GetFullPath(System.IO.Path.Combine(BNFTPPath, Name))
            End Get
        End Property

        Public Property StreamReader As StreamReader = Nothing

        Public Sub New(ByVal filename As String)
            Name = filename
        End Sub

        Public Sub Close()
            CloseStream()
        End Sub

        Public Sub CloseStream()
            If StreamReader IsNot Nothing Then
                StreamReader.Close()
            End If
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            Close()
        End Sub

        Public Function GetFileInfo(ByVal Optional ignoreLimits As Boolean = False) As FileInfo
            Dim rootStr = System.IO.Path.GetFullPath(BNFTPPath)
            Dim pathStr = System.IO.Path.GetFullPath(System.IO.Path.Combine(rootStr, Name))

            If Not ignoreLimits AndAlso (pathStr.Length < rootStr.Length OrElse pathStr.Substring(0, rootStr.Length) <> rootStr) Then
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info; path would leave BNFTP root directory")
                Return Nothing
            End If

            Dim fileinfo As FileInfo

            Try
                fileinfo = New FileInfo(System.IO.Path.Combine(BNFTPPath, Name))
            Catch ex As Exception
                If Not (TypeOf ex Is UnauthorizedAccessException OrElse TypeOf ex Is PathTooLongException OrElse TypeOf ex Is NotSupportedException) Then Throw
                Return Nothing
            End Try

            If fileinfo Is Nothing Then
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; null FileInfo object")
                Return Nothing
            End If

            If Not ignoreLimits AndAlso Not fileinfo.Exists Then
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; file not found")
                Return Nothing
            End If

            If Not ignoreLimits AndAlso fileinfo.Attributes.HasFlag(FileAttributes.Directory) Then
                Logging.WriteLine(Logging.LogLevel.Warning, Logging.LogType.BNFTP, $"Error retrieving file info for [{Name}]; path pointed to a directory")
                Return Nothing
            End If

            Return fileinfo
        End Function

        Public Function OpenStream() As Boolean
            If StreamReader IsNot Nothing Then
                StreamReader.Close()
            End If

            StreamReader = Nothing
            Dim fileinfo = GetFileInfo()
            If fileinfo Is Nothing Then Return False
            Logging.WriteLine(Logging.LogLevel.Debug, Logging.LogType.BNFTP, $"Opening read stream for [{fileinfo.FullName}]...")
            StreamReader = New StreamReader(fileinfo.FullName)
            Return True
        End Function
    End Class
End Namespace